### PR TITLE
Enhance device capabilities and constraints management

### LINKED ISSUE #375: [TASK: ] Integrate Device Selector in WFM & WFM Client
### Description

# Task: Add Device Eligibility Evaluation to WFM CLI and WFM Client

## Description

Implement device selector evaluation in both the WFM CLI and WFM Client to support application deployment profiles containing `deviceConstraints`. This task includes exposing device eligibility information during device selection and enforcing eligibility validation on the target device before deployment. 【1-251985】

---

## Task 1: Add Device Eligibility Indicator to WFM CLI

Enhance the device listing functionality used during application deployment to evaluate the selected application's `deviceConstraints` against available devices and display the eligibility result.

### Requirements

- Evaluate application `deviceConstraints` for each device returned by the inventory service.
- Add an `Eligible` field to the device listing output.
- Populate the field with one of the following values:
  - `true` - Device satisfies the application's `deviceConstraints`.
  - `false` - Device does not satisfy the application's `deviceConstraints`.
  - `unknown` - Eligibility cannot be determined or evaluation is not applicable.
- Use the shared device selector implementation for eligibility evaluation.
- Preserve all existing device listing behavior. 【1-251985】

### Acceptance Criteria

- Device listings include an `Eligible` column.
- Eligibility is evaluated using application `deviceConstraints`.
- Eligibility values are displayed as `true`, `false`, or `unknown`.
- Existing CLI functionality remains unaffected other than the addition of the eligibility field.

---

## Task 2: Validate Device Eligibility in WFM Client

Enhance the WFM Client to evaluate incoming application deployment requests against its local device capabilities and labels before deployment processing begins.

### Requirements

- Evaluate deployment profile `deviceConstraints` against the local device capabilities.
- Use the shared device selector implementation for constraint validation.
- Proceed with deployment when the device satisfies all constraints.
- Reject deployment when the device does not satisfy the constraints.
- Record a clear failure reason indicating that device eligibility validation failed. 【1-251985】

### Status Handling

When eligibility validation fails:

- Set the application deployment status to **Failed**.
- Do **not** set the status to **Not Placeable**.

A constraint mismatch at the client occurs after a target device has already been selected. Therefore, the deployment should be treated as a deployment failure rather than a placement failure. 【1-251985】

### Acceptance Criteria

- WFM Client validates deployment profile `deviceConstraints` before deployment.
- Eligible deployments continue through the normal deployment workflow.
- Ineligible deployments transition to **Failed** status.
- Failure details clearly indicate a device eligibility validation failure.
- Deployments without `deviceConstraints` continue to follow existing behavior.

---

## Dependencies

- Shared device selector evaluation library.
- Device capabilities and label metadata required for selector evaluation. 【1-251985】

---

## Definition of Done

- Device eligibility is displayed in WFM CLI device listings.
- Device eligibility is evaluated using deployment profile `deviceConstraints`.
- WFM Client validates deployment eligibility before deployment execution.
- Ineligible deployments are marked as **Failed**.
- Unit and integration tests cover eligible, ineligible, and unknown eligibility scenarios.
- Relevant documentation is updated.

### Raised by (email-id)

pulkit.d.sharma@capgemini.com

### Type of Story

Implementation

### Related Tasks

https://github.com/margo/sandbox/issues/369

### Status

In Motion

### % Done

90

### Estimated Story-points

2

### Updated artifacts

- [ ] API Documentation (Swagger)
- [ ] User Docmentation
- [x] Reference Implementation Code
- [ ] Test Automation

### PR DESCRIPTION
**Pull request by spulkit138**
# Enhance device capabilities and constraints management

Summary
- This PR introduces a consolidated system for managing device capabilities and runtime constraints. Implementation of (Device Runtime SUP)[https://github.com/margo/specification-enhancements/blob/main/proposals/sup_device_specific_runtime_affinity_matching.md]
- Adds core logic to define, validate, store, and enforce capability definitions and constraint rules across devices.
- Includes implementation, tests, and documentation updates to support the new capabilities/constraints workflow.

Key changes
- New modules and/or services for:
  - Device capability model: canonical representation of device capabilities and metadata.
  - Constraints manager/engine: declarative constraint definitions, evaluation, and enforcement.
  - Validation and registration hooks: capability/constraint validation at device onboarding and updates.

- Persistence and migrations:
  - Adds persistency support for capability definitions and constraints (DB schema and/or storage updates included).
- Tests:
  - Unit tests and integration tests covering capability validation, constraint evaluation, edge cases and error handling.
- Docs:
  - Developer docs and usage examples for defining capabilities and constraints, and running migration/scripts.

Checklist
- [x] Implementation of device capability model
- [x] Constraints manager/engine implemented
- [x] Unit + integration tests added
- [x] Docs / usage examples added

### CODE REVIEW

File: docs/repo-structure.md [hunk 1]
```diff
@@ -100,12 +100,16 @@ Reusable Go libraries providing common functionality across MARGO components.
 - **Workload management** (`workloads/`) - Helm and Docker Compose clients
 - **Archive** (`archive/`) - Unpacking or packing archives(tar.gz) etc..
 - **Pointer operations** (`pointers/`) - Some helper functions to deep clone, safely get pointer to temp variables etc...
+- **Device Constraint Selector Engine** (`constraints/`) - Reusable device eligibility checking library for the checks defined in [Device Runtime Affinity SUP](https://github.com/margo/specification-enhancements/blob/main/proposals/sup_device_specific_runtime_affinity_matching.md)
```
Comment (vireshnavalli) [inline docs/repo-structure.md:103]: Need to change this path once it is merged to completed folder or better to provide `specification` PR path as there might be some changes post `specification-enhancement` PR approval.

File: docs/repo-structure.md [hunk 2]
```diff
@@ -100,12 +100,16 @@ Reusable Go libraries providing common functionality across MARGO components.
...
+- **Go Set Library** (`set/`) - A general purpose set implementation for Go, with commonly used operations in Go.
+- **IEC Quantity Parser** (`quantity/`) - A general purpose minimalist IEC mini quantity notation parser & comparer. 
 
 ### 🛠️ Development Tools (`scripts/`)
 Scripts and utilities for development, testing, and deployment automation.
 
 **Tools:**
 - **Setup script** (`wfm.sh`, `device-agent.sh`) - Automated environment setup (Harbor, device's Workload Fleet Management Client, Symphony etc.)
+- **Label Generator Script** (`create-device-labels.sh`) - This script helps users generate labels for a device as per the guidelines defined here: [Device Runtime Affinity SUP](https://github.com/margo/specification-enhancements/blob/main/proposals/sup_device_specific_runtime_affinity_matching.md)
```
Comment (vireshnavalli) [inline docs/repo-structure.md:112]: here are refer the `specification` PR because `specification-enhancement` moved to completed folder from `proposal` folder once spec PR is approved/merged to pre-draft

File: non-standard/spec/wfm-nbi.yaml [hunk 3]
```diff
@@ -381,6 +381,11 @@ paths:
           in: query
           schema: { type: string }
           description: Token for pagination
+        - name: appPackageId
```
Comment (vireshnavalli) [inline non-standard/spec/wfm-nbi.yaml:384]: what is the purpose of this additional parameter, where this input come from?

File: non-standard/spec/wfm-nbi.yaml [hunk 4]
```diff
@@ -641,6 +646,7 @@ components:
             oneOf:
             - $ref: '#/components/schemas/HelmApplicationDeploymentProfileComponent'
             - $ref: '#/components/schemas/ComposeApplicationDeploymentProfileComponent'
+            - $ref: '#/components/schemas/CustomApplicationDeploymentProfileComponent'
```
Comment (vireshnavalli) [inline non-standard/spec/wfm-nbi.yaml:649]: I think we have discussed this already, as you have one more profile component (`custom`). shall we merge all as single for better clarity? what is your thought?

File: poc/device/agent/config/capabilities.json [hunk 5]
```diff
@@ -1,19 +1,30 @@
...
     "apiVersion": "device.margo.org/v1alpha1",
     "kind": "DeviceCapabilitiesManifest",
+    "labels": {
+        "northstarida.com/hypervisor": "hyper-v",
+        "northstarida.com/wasm.runtime": [
+            "wamr"
+        ],
+        "northstarida.com/wasm.package.format": [
+            ".wasm",
+            ".aot"
```
Comment (vireshnavalli) [inline poc/device/agent/config/capabilities.json:11]: we have single static capability file for both `helm` and `compose` and we update supportedDeploymentTypes (any other attributes) during setup

Introducing these labels are not necessary from helm and compose right?

File: poc/device/agent/deployment.go [hunk 6]
```diff
@@ -138,6 +142,7 @@ func (dm *DeploymentManager) reconcileDeployment(deploymentId string) {
 		// Only deploy if not already installed
 		if currentState != sbi.DeploymentStatusManifestStatusStateInstalled {
 			dm.log.Debugw("deploying pending deployment", "deploymentId", deploymentId)
+			// TODO: Check if device is eligible or not, if not, set it to failed state even before deployment
```
Comment (vireshnavalli) [inline poc/device/agent/deployment.go]: is checking device eligibility planned later? IMO, WFM will filter device is eligible or not.?

File: docs/repo-structure.md (hunk 1 above)
Comment (spulkit138) [reply docs/repo-structure.md:103]: I checked in [specification PR](https://github.com/margo/specification/pull/203/) , this exact content is not there, hence I would say we can change path once it is moved to `completed` folder.

File: docs/repo-structure.md (hunk 2 above)
Comment (spulkit138) [reply docs/repo-structure.md:112]: I checked in https://github.com/margo/specification/pull/203/ , this exact content is not there, hence I would say we can change path once it is moved to `completed` folder, as this PR will be merged in `main` when specification PR will be merged.

File: non-standard/spec/wfm-nbi.yaml (hunk 3 above)
Comment (spulkit138) [reply non-standard/spec/wfm-nbi.yaml:384]: As discussed, WFM CLI would show "eligible or not eligible" while selecting device for a particular deployment. For that eligibility check, `List Device` needs an optional parameter "application package Id" in order to get application device constraints for that particular application package. More information is present in description of this PR : https://github.com/margo/sandbox/pull/374

File: non-standard/spec/wfm-nbi.yaml (hunk 4 above)
Comment (spulkit138) [reply non-standard/spec/wfm-nbi.yaml:649]: I pointed out this previously that does non-standard API need to change as well in compose oci changes. My point was, for parts which are same in standard & non-standard, we should keep standard as single source of truth. Hence change here should be done and should be same to what is present in standard. Custom Runtime PR introduces a `custom` runtime apart from helm/compose. Hence this change was added. 

If we align to change it, my suggestion would be to either change it as a separate PR and then we pull changes in this to align it; OR after merging this a separate PR should be raised for that change. That ways, we can keep these two things separate.

File: poc/device/agent/config/capabilities.json (hunk 5 above)
Comment (spulkit138) [reply poc/device/agent/config/capabilities.json:11]: Labels are totally optional. Here, they are just placeholders for reference. They are not used while deployment. In installation scripts, I have provided a different way to add labels while installing WFM Client. And if labels are not provided at installation time, these placeholder labels are removed automatically. 

Keeping it here just because if someone wants to deploy as a binary, they will know how labels can be added, and they can modify/remove as required.

File: poc/device/agent/deployment.go (hunk 6 above)
Comment (spulkit138) [reply poc/device/agent/deployment.go]: I forgot to remove this TODO comment. I have removed it now. 

As per discussion, WFM will just show whether a device is eligible or not. Selection of a particular device still remains with the user. However, on Device agent side, if a particular application is ineligible, it is put in FAILED mode without even deploying it.
